ksql-streams/src/main/java/io/confluent/ksql/execution/streams/SinkBuilder.java

```diff
@@ -0,0 +1,132 @@
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.streams;
+
+import static java.util.Objects.requireNonNull;
+
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.execution.builder.KsqlQueryBuilder;
+import io.confluent.ksql.execution.context.QueryContext;
+import io.confluent.ksql.execution.plan.Formats;
+import io.confluent.ksql.execution.plan.KeySerdeFactory;
+import io.confluent.ksql.execution.timestamp.TimestampColumn;
+import io.confluent.ksql.schema.ksql.Column;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.PhysicalSchema;
+
+import java.util.Objects;
+import java.util.Optional;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.Produced;
+import org.apache.kafka.streams.kstream.Transformer;
+import org.apache.kafka.streams.kstream.TransformerSupplier;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.To;
+
+public final class SinkBuilder {
+  private SinkBuilder() {
+  }
+
+  public static  <K> void build(
+      final LogicalSchema schema,
+      final Formats formats,
+      final Optional<TimestampColumn> timestampColumn,
+      final String topicName,
+      final KStream<K, GenericRow> stream,
+      final KeySerdeFactory<K> keySerdeFactory,
+      final QueryContext queryContext,
+      final KsqlQueryBuilder queryBuilder
+  ) {
+    final PhysicalSchema physicalSchema = PhysicalSchema.from(schema, formats.getOptions());
+
+    final Serde<K> keySerde = keySerdeFactory.buildKeySerde(
+        formats.getKeyFormat(),
+        physicalSchema,
+        queryContext
+    );
+
+    final Serde<GenericRow> valueSerde = queryBuilder.buildValueSerde(
+        formats.getValueFormat(),
+        physicalSchema,
+        queryContext
+    );
+
+    final int timestampColumnIndex = timestampColumn.map(TimestampColumn::getColumn)
+        .map(c -> schema.findValueColumn(c).orElseThrow(IllegalStateException::new))
+        .map(Column::index)
+        .orElse(-1);
+
+    stream.transform(new TransformTimestamp<>(timestampColumnIndex))
+        .to(topicName, Produced.with(keySerde, valueSerde));
+  }
+
+  static class TransformTimestamp<K>
+      implements TransformerSupplier<K, GenericRow, KeyValue<K, GenericRow>> {
+    private final int timestampColumnIndex;
+
+    TransformTimestamp(final int timestampColumnIndex) {
+      this.timestampColumnIndex = timestampColumnIndex;
```

Comment:
nit: validate params that will be stored in object state; ensuring object does not get into an invalid state, i.e. throw if negative.

Reply:
Done